packages/jupyter-ai/jupyter_ai/extension.py

```diff
@@ -204,6 +234,127 @@ class AiExtension(ExtensionApp):
         config=True,
     )
 
+    def initialize(self):
+        super().initialize()
+        self.event_logger = self.settings["event_logger"]
+        self.event_logger.add_listener(
+            schema_id=JUPYTER_COLLABORATION_EVENTS_URI, listener=self.connect_chat
+        )
+
+        # Keep the message indexes to avoid extra computation looking for a message when
+        # updating it.
+        self.messages_indexes = {}
+
+    async def connect_chat(
+        self, logger: EventLogger, schema_id: str, data: dict
+    ) -> None:
+        if (
+            data["room"].startswith("text:chat:")
+            and data["action"] == "initialize"
+            and data["msg"] == "Room initialized"
+        ):
+
+            self.log.info(f"Collaborative chat server is listening for {data['room']}")
+            chat = await self.get_chat(data["room"])
+
+            if chat is None:
+                return
+
+            # Add the bot user to the chat document awareness.
+            BOT["avatar_url"] = url_path_join(
+                self.settings.get("base_url", "/"), "api/ai/static/jupyternaut.svg"
+            )
+            if chat.awareness is not None:
+                chat.awareness.set_local_state_field("user", BOT)
+
+            callback = partial(self.on_change, chat)
+            chat.ymessages.observe(callback)
+
+    async def get_chat(self, room_id: str) -> Optional[YChat]:
+        if COLLAB_VERSION == 3:
+            collaboration = self.settings["jupyter_server_ydoc"]
+            document = await collaboration.get_document(room_id=room_id, copy=False)
+        else:
+            collaboration = self.settings["jupyter_collaboration"]
+            server = collaboration.ywebsocket_server
+
+            room = await server.get_room(room_id)
+            document = room._document
+        return document
+
+    def on_change(self, chat: YChat, events: ArrayEvent) -> None:
+        for change in events.delta:  # type:ignore[attr-defined]
+            if not "insert" in change.keys():
+                continue
+            messages = change["insert"]
+            for message in messages:
+
+                if message["sender"] == BOT["username"] or message["raw_time"]:
+                    continue
```

Comment:
> * `jupyterlab-chat` needs to have an attribute on the client model which describes if a client is a human user or another bot.

Not sure about including the notion of human or bot in `jupyterlab-chat`, this would not be used there. The user model (only described in [javascript](https://github.com/jupyterlab/jupyter-chat/blob/967df3054b305275163a5c7d0de47a804b90471c/packages/jupyter-chat/src/types.ts#L9-L16) currently) reflect the one described in jupyter-server and jupyterlab.
Maybe we could add this attribute from `jupyyer-ai`.

> * The Python API should only provide a way for server extensions to add a bot, such that other server extensions don't accidentally/maliciously register a bot appearing as a human user.

A bot could also be registered from the frontend.

> * This code should be modified to skip non-human users, i.e. Jupyternaut should only reply to queries from human users and never reply to other bots.

Agree

Reply:
Let's chat about this more later. I'm tracking this in my personal notes, which I will document in future issues.